### PR TITLE
Request cancel

### PR DESCRIPTION
**Pull request by bdunne**
Based on changes in: https://github.com/ManageIQ/manageiq/pull/17687
https://bugzilla.redhat.com/show_bug.cgi?id=1564257
cc @abellotti @gtanzillo @gmcculloug @bzwei

### CODE REVIEW

**Comment by bzwei** [comment]
@bdunne v2v RFE needs task level cancellation. Do we expect another PR?

**Comment by gmcculloug** [comment]
@bdunne Since the back-end is really only supporting this feature for v2v can we reduce the models this is exposed on?  At the moment we return "Unsupported" for all other request/task types so I would prefer limiting the action as much as possible.

**Comment by bdunne** [comment]
@gmcculloug Many thoughts..
- I was surprised to find out that we have 4 requests endpoints.  Do you know why?
- I'd like keep these 4 endpoints as similar as possible as long as it makes sense (especially in this case where 2/4 would have to implement it in my opinion)
- I think an "Unsupported" response is better than "Bad Request Error"
- It may only be used by v2v at the moment, but I don't think the API should care about implementing things for specific use cases, it should be done in a generic and consistent way.

**Comment by JPrause** [comment]
@miq-bot add_label blocker

**Comment by JPrause** [comment]
@bdunne if this can be backported, can you add the gaprindashvili/yes label

**Comment by gmcculloug** [comment]
@gtanzillo @abellotti Any thoughts on https://github.com/ManageIQ/manageiq-api/pull/421#issuecomment-405581100

**Comment by gtanzillo** [comment]
@gmcculloug Regarding
>I was not involved in building out the API so do not have any insight into why there are 4 endpoints.

>@gtanzillo @abellotti Any thoughts on #421 (comment)

I was also not involved so I'm not sure either. It seems in some places we've exposed a single endpoint for the base class and in other places we exposed separate endpoints for each subclass like was done here. 

Perhaps @abellotti or @Fryguy can comment on that. It would be good to understand which convention we should be following, if there is one.

**Comment by bdunne** [comment]
Okay, I think this is ready for review.  @jrafanie can you put another set of :eyes: on the tests?

**Comment by jrafanie** [comment]
LGTM.  It's a bit hard to follow at first with the indirection but the alternative is 1000 lines of duplicated tests, so :shipit: 

It would be good if @Fryguy can comment on the separate collections for each subclass.  It seems strange.  I can't imagine having separate collections for vmware  vms, redhat vms, etc.   On the fipside, these things are not the same and are actual tasks for completely different end user features. 🤷‍♂️

File: app/controllers/api/mixins/resource_cancel.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module Api
+  module Mixins
+    module ResourceCancel
```
Comment (Fryguy) [inline app/controllers/api/mixins/resource_cancel.rb]: Is this ResourceCancel or TaskCancel?

**Comment by Fryguy** [comment]
> It would be good if @Fryguy can comment on the separate collections for each subclass. It seems strange. I can't imagine having separate collections for vmware vms, redhat vms, etc.

Where are you seeing this?

File: spec/support/api/helpers.rb [hunk 2]
```diff
@@ -224,6 +224,12 @@ def expect_options_results(type, data = {})
           expect(response.headers['Access-Control-Allow-Methods']).to include('OPTIONS')
         end
 
+        def expect_unauthorized_request
```
Comment (Fryguy) [inline spec/support/api/helpers.rb]: minor, but this might read better as expect_forbidden_request, since the status is `:forbidden`

File: app/controllers/api/mixins/resource_cancel.rb (hunk 1 above)
Comment (bdunne) [reply app/controllers/api/mixins/resource_cancel.rb]: It's for `ServiceOrder`, `MiqRequest`X4 or `MiqRequestTask` primary collections.

**Comment by bdunne** [comment]
> @jrafanie Where are you seeing this?

`/api/requests`
`/api/automation_requests`
`/api/service_requests`
`/api/provision_requests`

Talking to Alberto, it sounds like it just grew this way over time.  Hopefully we don't need to create new endpoints for the other 12 subclasses of `MiqRequest`.

**Comment by AparnaKarve** [comment]
Can we get this merged today?

**Comment by bthurber** [comment]
@gtanzillo @gmcculloug @blomquisg we will need this merged ASAP for 5.9.4-2 build for tomorrow.  Anything here left to finish/review?

**Comment by bthurber** [comment]
@bdunne can you respond to @Fryguy last comment?  We also need to add enhancement, transformation and grapendashvili/yes labels so this get's pulled into the next build once merged.

**Comment by bdunne** [comment]
@bthurber I responded in: https://github.com/ManageIQ/manageiq-api/pull/421#issuecomment-406416017

**Comment by bdunne** [comment]
@miq-bot add_labels gaprindashvili/yes enhancement transformation

**Comment by bdunne** [comment]
@miq-bot add_labels gaprindashvili/yes
@miq-bot add_labels enhancement
@miq-bot add_labels transformation

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit b7047744b21596fafe4cbfdc1064cad1fd8ba76c
Author: Gregg Tanzillo <gtanzill@redhat.com>
Date:   Mon Jul 30 18:02:12 2018 -0400

    Merge pull request #421 from bdunne/request_cancel
    
    Request cancel
    (cherry picked from commit 62de390490c8bc333f27feedfc58455ced1e7483)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1610533
```
